crates/walrus-core/src/encoding/quilt_encoding.rs

```diff
@@ -0,0 +1,1367 @@
+}
+
+pub trait QuiltApi<V: QuiltVersion> {
+    /// Gets a blob by its identifier from the quilt.
+    fn get_blob_by_identifier(&self, identifier: &str) -> Result<Vec<u8>, QuiltError>;
+
+    /// Returns the quilt index.
+    fn quilt_index(&self) -> &V::QuiltIndex;
+
+    /// Returns the data of the quilt.
+    fn data(&self) -> &[u8];
+
+    /// Returns the symbol size of the quilt.
+    fn symbol_size(&self) -> usize;
+}
+
+pub struct QuiltVersionV1;
+
+impl QuiltVersion for QuiltVersionV1 {
+    type QuiltConfig = QuiltConfigV1;
+    type QuiltEncoder<'a> = QuiltEncoderV1<'a>;
+    type QuiltDecoder<'a> = QuiltDecoderV1<'a>;
+    type Quilt = QuiltV1;
+    type QuiltIndex = QuiltIndexV1;
+    type QuiltMetadata = QuiltMetadata;
+}
+
+/// The configuration of a quilt.
+#[derive(Serialize, Deserialize, PartialEq, Eq, Debug)]
+pub struct QuiltConfigV1 {}
+
+impl<'a> QuiltConfigApi<'a, QuiltVersionV1> for QuiltConfigV1 {
+    fn get_encoder(
+        &self,
+        encoding_config: EncodingConfigEnum<'a>,
+        blobs: &'a [BlobWithIdentifier<'a>],
+    ) -> QuiltEncoderV1<'a> {
+        QuiltEncoderV1::new(encoding_config, blobs)
+    }
+
+    fn get_decoder(&self, slivers: &'a [&'a SliverData<Secondary>]) -> QuiltDecoderV1<'a> {
+        QuiltDecoderV1::new(slivers)
+    }
+
+    fn new_from_quilt_blob(
+        &self,
+        quilt_blob: Vec<u8>,
+        metadata: &QuiltMetadata,
+        n_shards: NonZeroU16,
+    ) -> Result<QuiltV1, QuiltError> {
+        // assert_eq!(self, &QuiltConfig::V1);
+        let encoding_config = EncodingConfig::new(n_shards);
+        let config = encoding_config.get_for_type(metadata.metadata().encoding_type());
+
+        let n_primary_source_symbols = config.n_primary_source_symbols().get();
+        let n_secondary_source_symbols = config.n_secondary_source_symbols().get();
+        let n_source_symbols = n_primary_source_symbols * n_secondary_source_symbols;
+
+        // Verify data alignment.
+        if quilt_blob.len() % n_source_symbols as usize != 0 {
+            return Err(QuiltError::invalid_format_not_aligned());
+        }
+
+        // Calculate matrix dimensions.
+        let row_size = quilt_blob.len() / n_primary_source_symbols as usize;
+        let symbol_size = row_size / n_secondary_source_symbols as usize;
+
+        // Extract quilt index size and calculate required columns.
+        let quilt_index = utils::get_quilt_index_from_data(&quilt_blob, row_size, symbol_size)?;
+        assert_eq!(quilt_index, metadata.index);
+
+        Ok(QuiltV1 {
+            data: quilt_blob,
+            row_size,
+            quilt_index,
+            symbol_size,
+        })
+    }
+}
+
+/// A quilt is a collection of encoded blobs stored together in a unified structure.
+///
+/// The data is organized as a 2D matrix where:
+/// - Each blob occupies a continuous range of columns (secondary slivers).
+/// - The first column's initial `QUILT_INDEX_SIZE_PREFIX_SIZE` bytes contain the unencoded
```

Comment:
Suggestion: Explicitly include the quilt version at the very beginning of the raw data. That way we wouldn't need any additional information to parse a quilt.